rsciio/topspin/_api.py

```diff
@@ -0,0 +1,360 @@
+
+import sys
+import xml.etree.cElementTree as ET
+
+import h5py
+import numpy as np
+from tqdm import tqdm
+
+from rsciio._docstrings import FILENAME_DOC, RETURNS_DOC, SHOW_PROGRESSBAR_DOC
+
+
+def file_reader(filename, subset=None, dryrun=False, show_progressbar=True):
+    """
+    Read .app5 file format used by NanoMegas's Topspin software.
+
+    .app5 files use the hdf5 file format, with metadata stored as
+    binarized XML-style text strings.
+
+
+    Parameters
+    ----------
+    %s
+    subset: str or None
+        h5py-style address. If given, only the subset of experiments
+        located at this address will be imported. If none, all datasets
+        will be imported.
+
+    dryrun : bool
+        If True, the .app5 files are quickly scanned without being loaded,
+        and a summary is printed to the log. Default is False.
+
+    %s
+    %s
+
+    Notes
+    -----
+    The Metadata textstrings used in app5 files change based on the
+    ProcedureName, Topspin version, and local microscope setup. Because
+    of this, the RosettaSciIo Metadata parser can fail when used to
+    decode newer Topspin proceedures it was not tested against.
+    """
+
+    def is_guid(name):
+        """Checks if a text string matches the shape of a guid"""
+        lengths = [len(x) for x in name.split("-")]
+        return lengths == [8, 4, 4, 4, 12]
+
+    datasets_list = []
+    task_list = []
+    h5_file = h5py.File(filename, "r")
+    if subset is not None:
+        h5_file = h5_file[subset]
+
+    # Generate task list by searching .app5 for GUID's
+    group_names = [x for x in h5_file.keys()]
+    if np.isin("Metadata", group_names):  # single experiment file.
+        for name in h5_file.keys():
+            if is_guid(name):
+                task_list.append([name, "Metadata"])
+    else:  # Multi-experimental file.
+        for top_name in h5_file.keys():
+            if is_guid(top_name):
+                for name in h5_file[top_name].keys():
+                    if is_guid(name):
+                        address = top_name + "/" + name
+                        meta = top_name + "/" + "Metadata"
+                        task_list.append([address, meta])
+
+    # Parse each dataset's metadata
+    for address, meta_loc in task_list:
+        name = address.split("/")[-1]
+        metadata_string = h5_file[meta_loc][()].decode()
+        meta_dict = _parse_app5_xml(metadata_string)
+        if isinstance(h5_file[address], h5py.Group):
+            # 4D-STEM
+            h5_grp = h5_file[address]
+            axes = _get_4D_axes(h5_grp)
+        else:
+            # 2D composite image
+            axes = _get_2D_axes(metadata_string, name)
+            shape = h5_file[address].shape
+            axes[0]["size"] = shape[0]
+            axes[1]["size"] = shape[1]
+        datasets_list.append(
+            {
+                "axes": axes,
+                "metadata": meta_dict,
```

Comment:
`meta_dict` (all metadata parsed as python dictionary) go to `original_metadata`. `metadata` is for metadata defined in https://hyperspy.org/hyperspy-doc/current/reference/metadata.html or other extension metadata (https://docs.lumispy.org/en/stable/user_guide/metadata_structure.html, https://hyperspy.org/exspy/user_guide/metadata_structure.html, pyxem must be using some too, but I don't know where is documented).